File: structural/dart/decorator_structured/decorators.dart
import 'components.dart';

abstract class CoffeeDecorator extends Coffee {
  final Coffee decoratedCoffee;
  CoffeeDecorator(this.decoratedCoffee);
  @override
  double getCost() { return decoratedCoffee.getCost(); }
  @override
  String getDescription() { return decoratedCoffee.getDescription(); }
}

class MilkDecorator extends CoffeeDecorator {
  MilkDecorator(super.coffee);
  @override
  double getCost() { return super.getCost() + 0.5; }
  @override
  String getDescription() { return '\${super.getDescription()}, Milk'; }
}

class SugarDecorator extends CoffeeDecorator {
  SugarDecorator(super.coffee);
  @override
  double getCost() { return super.getCost() + 0.2; }
  @override
  String getDescription() { return '\${super.getDescription()}, Sugar'; }
}

class WhippedCreamDecorator extends CoffeeDecorator {
  WhippedCreamDecorator(super.coffee);
  @override
  double getCost() { return super.getCost() + 1.0; }
  @override
  String getDescription() { return '\${super.getDescription()}, Whipped Cream'; }
}

abstract class TextDecorator extends TextComponent {
  final TextComponent wrappedText;
  TextDecorator(this.wrappedText);
  @override
  String getText() { return wrappedText.getText(); }
}

class BoldDecorator extends TextDecorator {
  BoldDecorator(super.text);
  @override
  String getText() { return '<b>\${super.getText()}</b>'; }
}

class ItalicDecorator extends TextDecorator {
  ItalicDecorator(super.text);
  @override
  String getText() { return '<i>\${super.getText()}</i>'; }
}

class UnderlineDecorator extends TextDecorator {
  UnderlineDecorator(super.text);
  @override
  String getText() { return '<u>\${super.getText()}</u>'; }
}

abstract class DataSourceDecorator extends DataSource {
  final DataSource wrappedSource;
  DataSourceDecorator(this.wrappedSource);
  @override
  void writeData(String data) { wrappedSource.writeData(data); }
  @override
  String readData() { return wrappedSource.readData(); }
}

class EncryptionDecorator extends DataSourceDecorator {
  EncryptionDecorator(super.source);
  @override
  void writeData(String data) {
    var encrypted = '[ENCRYPTED]\$data[ENCRYPTED]';
    print('Encrypting data...');
    wrappedSource.writeData(encrypted);
  }
  @override
  String readData() {
    var data = wrappedSource.readData();
    print('Decrypting data...');
    if (data.startsWith('[ENCRYPTED]')) {
      return data.substring(11, data.length - 11);
    }
    return data;
  }
}

class CompressionDecorator extends DataSourceDecorator {
  CompressionDecorator(super.source);
  @override
  void writeData(String data) {
    var compressed = '[COMPRESSED]\$data[COMPRESSED]';
    print('Compressing data...');
    wrappedSource.writeData(compressed);
  }
  @override
  String readData() {
    var data = wrappedSource.readData();
    print('Decompressing data...');
    if (data.startsWith('[COMPRESSED]')) {
      return data.substring(12, data.length - 12);
    }
    return data;
  }
}

File: structural/typescript/decorator_structured/decorators.ts
import { Coffee, TextComponent, DataSource } from './components';

export abstract class CoffeeDecorator implements Coffee {
  protected decoratedCoffee: Coffee;
  constructor(coffee: Coffee) { this.decoratedCoffee = coffee; }
  getCost(): number { return this.decoratedCoffee.getCost(); }
  getDescription(): string { return this.decoratedCoffee.getDescription(); }
}

export class MilkDecorator extends CoffeeDecorator {
  getCost(): number { return super.getCost() + 0.5; }
  getDescription(): string { return super.getDescription() + ', Milk'; }
}

export class SugarDecorator extends CoffeeDecorator {
  getCost(): number { return super.getCost() + 0.2; }
  getDescription(): string { return super.getDescription() + ', Sugar'; }
}

export class WhippedCreamDecorator extends CoffeeDecorator {
  getCost(): number { return super.getCost() + 1.0; }
  getDescription(): string { return super.getDescription() + ', Whipped Cream'; }
}

export abstract class TextDecorator implements TextComponent {
  protected wrappedText: TextComponent;
  constructor(text: TextComponent) { this.wrappedText = text; }
  getText(): string { return this.wrappedText.getText(); }
}

export class BoldDecorator extends TextDecorator {
  getText(): string { return `<b>\${super.getText()}</b>`; }
}

export class ItalicDecorator extends TextDecorator {
  getText(): string { return `<i>\${super.getText()}</i>`; }
}

export class UnderlineDecorator extends TextDecorator {
  getText(): string { return `<u>\${super.getText()}</u>`; }
}

export abstract class DataSourceDecorator implements DataSource {
  protected wrappedSource: DataSource;
  constructor(source: DataSource) { this.wrappedSource = source; }
  writeData(data: string): void { this.wrappedSource.writeData(data); }
  readData(): string { return this.wrappedSource.readData(); }
}

export class EncryptionDecorator extends DataSourceDecorator {
  writeData(data: string): void {
    const encrypted = `[ENCRYPTED]\${data}[ENCRYPTED]`;
    console.log('Encrypting data...');
    this.wrappedSource.writeData(encrypted);
  }
  readData(): string {
    const data = this.wrappedSource.readData();
    console.log('Decrypting data...');
    if (data.startsWith('[ENCRYPTED]')) {
      return data.substring(11, data.length - 11);
    }
    return data;
  }
}

export class CompressionDecorator extends DataSourceDecorator {
  writeData(data: string): void {
    const compressed = `[COMPRESSED]\${data}[COMPRESSED]`;
    console.log('Compressing data...');
    this.wrappedSource.writeData(compressed);
  }
  readData(): string {
    const data = this.wrappedSource.readData();
    console.log('Decompressing data...');
    if (data.startsWith('[COMPRESSED]')) {
      return data.substring(12, data.length - 12);
    }
    return data;
  }
}
